syncthing/syncthing
# Add WebAuthn support to GUI

Fixes #8409. Replaces PR #8417.

### Remaining things to do

- [ ] Open documentation PR

### Purpose

This adds support for [WebAuthn](https://webauthn.guide/) authentication in the GUI. WebAuthn is a W3C standard that provides an API for strong, privacy-conscious public-key authentication, which can enable a both more pleasant and more secure user experience than traditional password authentication. WebAuthn both works with external security keys (typically via USB or NFC) and is built into many modern browsers and operating systems.

For users who already use WebAuthn, this both cuts out a detour through a password manager and improves security by making it possible to not need a password at all.

I'm opening this as a draft PR so people can try it out while I work on finishing the last few things listed above.


### Testing

Manual tests performed:

- With no password set and no WebAuthn credentials enrolled:
  - No authentication is required

- With password set OR at least one WebAuthn credential enrolled:
  - "Log in with WebAuthn" button added to login page
  - With no WebAuthn credentials enrolled, WebAuthn button is disabled
  - Password authentication works (no WebAuthn required)
  - WebAuthn authentication works (no password required)
  - WebAuthn authentication is available immediately without server restart

- WebAuthn settings/credential management:
  - Warning appears when HTTPS is disabled
  - Warning appears when RP ID does not match page domain
  - Warning appears when WebAuthn origin does not match RP ID
  - Warning appears when changing RP ID
  - New credentials appear immediately in the table in GUI settings, but take effect on saving the GUI settings
  - Renaming credentials takes effect on saving the GUI settings
  - Credential names fall back to credential ID when not set
  - Deleting credentials takes effect on saving the GUI settings
  - `PUT /rest/config` ignores WebAuthn credentials not already registered (identified by ID)
  - [`excludeCredentials`][exclude] is used to prevent registering the same authenticator twice
  - When "Require UV" is checked for a credential, user verification (PIN or biometric) is required when using that credential (hacking the frontend to override the authentication arguments with `userVerification = "discouraged"` fails).

I intend to add automated tests as well, but would first like a review of the overall design and how it fits into the existing architecture.


### Screenshots

![WebAuthn button added to login form](https://github.com/syncthing/syncthing/assets/1367758/07c9ec4c-5921-461d-b395-2ca12bf3f531)
WebAuthn button added to login form

![New GUI settings](https://github.com/syncthing/syncthing/assets/1367758/ac0158b6-b569-4bb1-aea0-84feb4bbf564)
New GUI settings

![Registering a new credential](https://github.com/syncthing/syncthing/assets/1367758/93536a59-e5b1-4266-b6af-f03c5568a7d5)
Registering a new credential

![Renaming a credential](https://github.com/syncthing/syncthing/assets/1367758/53ae9b79-e957-4de0-a136-6f781196c394)
Renaming a credential

![Interactive warnings](https://github.com/syncthing/syncthing/assets/1367758/476e0bda-d6f2-43a1-bca1-7b99c363fb36)
Interactive warnings


### Documentation

Probably a new section should be added alongside "LDAP Authentication". WebAuthn normally doesn't require much technical knowledge from the user since you only interact with the end-user side, but in the Syncthing case you are both end-user and server administrator. WebAuthn requires some domain settings to line up, so the "RP ID" and "WebAuthn Origin" settings and their impact need to be explained here. I volunteer to write this documentation once I have the green light to finish the feature.


[exclude]: https://w3c.github.io/webauthn/#dom-publickeycredentialcreationoptions-excludecredentials
[transports]: https://w3c.github.io/webauthn/#dom-publickeycredentialdescriptor-transports
[signCount]: https://w3c.github.io/webauthn/#signcount
